crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,542 @@
 // Pipewire backend device
 // SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
 
-use std::sync::{Arc, RwLock};
+use crate::device::ControlMessage;
+use crate::Result;
+use virtio_queue::Descriptor;
+use vm_memory::Bytes;
 
 use super::AudioBackend;
-use crate::{Result, Stream};
+use crate::stream::NR_STREAMS;
+use crate::virtio_sound::{
+    VirtioSndPcmSetParams, VIRTIO_SND_D_INPUT, VIRTIO_SND_D_OUTPUT, VIRTIO_SND_PCM_FMT_A_LAW,
+    VIRTIO_SND_PCM_FMT_FLOAT, VIRTIO_SND_PCM_FMT_FLOAT64, VIRTIO_SND_PCM_FMT_MU_LAW,
+    VIRTIO_SND_PCM_FMT_S16, VIRTIO_SND_PCM_FMT_S18_3, VIRTIO_SND_PCM_FMT_S20,
+    VIRTIO_SND_PCM_FMT_S20_3, VIRTIO_SND_PCM_FMT_S24, VIRTIO_SND_PCM_FMT_S24_3,
+    VIRTIO_SND_PCM_FMT_S32, VIRTIO_SND_PCM_FMT_S8, VIRTIO_SND_PCM_FMT_U16,
+    VIRTIO_SND_PCM_FMT_U18_3, VIRTIO_SND_PCM_FMT_U20, VIRTIO_SND_PCM_FMT_U20_3,
+    VIRTIO_SND_PCM_FMT_U24, VIRTIO_SND_PCM_FMT_U24_3, VIRTIO_SND_PCM_FMT_U32,
+    VIRTIO_SND_PCM_FMT_U8, VIRTIO_SND_PCM_RATE_11025, VIRTIO_SND_PCM_RATE_16000,
+    VIRTIO_SND_PCM_RATE_176400, VIRTIO_SND_PCM_RATE_192000, VIRTIO_SND_PCM_RATE_22050,
+    VIRTIO_SND_PCM_RATE_32000, VIRTIO_SND_PCM_RATE_384000, VIRTIO_SND_PCM_RATE_44100,
+    VIRTIO_SND_PCM_RATE_48000, VIRTIO_SND_PCM_RATE_5512, VIRTIO_SND_PCM_RATE_64000,
+    VIRTIO_SND_PCM_RATE_8000, VIRTIO_SND_PCM_RATE_88200, VIRTIO_SND_PCM_RATE_96000,
+    VIRTIO_SND_S_BAD_MSG, VIRTIO_SND_S_NOT_SUPP,
+};
+use crate::{Error, Stream};
+use std::{
+    cmp,
+    collections::HashMap,
+    convert::TryInto,
+    mem::size_of,
+    ops::Deref,
+    ptr,
+    ptr::NonNull,
+    sync::{Arc, RwLock},
+};
+
+use log::debug;
+use pw::spa::{param::ParamType, pod::Pod};
+
+use pw::spa::sys::{
```

Comment:
you already do use `pw::spa` so you can drop the `pw` prefix here
```suggestion
use spa::sys::{
```